src/main/java/graphql/util/FpKit.java

```diff
@@ -18,69 +19,87 @@
     }
 
     // normal groupingBy but with LinkedHashMap
     public static <T, NewKey> Map<NewKey, ImmutableList<T>> groupingBy(Collection<T> list, Function<T, NewKey> function) {
-        return list.stream().collect(Collectors.groupingBy(function, LinkedHashMap::new, mapping(Function.identity(), ImmutableList.toImmutableList())));
+        return filterAndGroupingBy(list, t -> true, function);
     }
 
+    @SuppressWarnings("unchecked")
     public static <T, NewKey> Map<NewKey, ImmutableList<T>> filterAndGroupingBy(Collection<T> list,
                                                                                 Predicate<? super T> predicate,
                                                                                 Function<T, NewKey> function) {
-        return list.stream().filter(predicate).collect(Collectors.groupingBy(function, LinkedHashMap::new, mapping(Function.identity(), ImmutableList.toImmutableList())));
-    }
+        //
+        // The cleanest version of this code would have two maps, one of immutable list builders and one
+        // of the built immutable lists.  BUt we are trying to be performant and memory efficient so
+        // we treat it as a map of objects and cast like its Java 4x
+        //
+        Map<NewKey, Object> resutMap = new LinkedHashMap<>();
+        for (T item : list) {
+            if (predicate.test(item)) {
+                NewKey key = function.apply(item);
+                // we have to use an immutable list builder as we built it out
+                ((ImmutableList.Builder<Object>) resutMap.computeIfAbsent(key, k -> ImmutableList.builder()))
+                        .add(item);
+            }
+        }
+        if (resutMap.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        // Convert builders to ImmutableLists in place to avoid an extra allocation
+        // yes the code is yuck -  but its more performant yuck!
+        resutMap.replaceAll((key, builder) ->
+                ((ImmutableList.Builder<Object>) builder).build());
 
-    public static <T, NewKey> Map<NewKey, ImmutableList<T>> groupingBy(Stream<T> stream, Function<T, NewKey> function) {
-        return stream.collect(Collectors.groupingBy(function, LinkedHashMap::new, mapping(Function.identity(), ImmutableList.toImmutableList())));
+        // make it the right shape - like as if generics were never invented
+        return (Map<NewKey, ImmutableList<T>>) (Map<?, ?>) resutMap;
     }
 
-    public static <T, NewKey> Map<NewKey, T> groupingByUniqueKey(Collection<T> list, Function<T, NewKey> keyFunction) {
-        return list.stream().collect(Collectors.toMap(
-                keyFunction,
-                identity(),
-                throwingMerger(),
-                LinkedHashMap::new)
-        );
+    public static <T, NewKey> Map<NewKey, T> toMapByUniqueKey(Collection<T> list, Function<T, NewKey> keyFunction) {
+        return toMap(list, keyFunction, throwingMerger());
     }
```

Comment:
This was renamed from `groupingByUniqueKey` to `toMapByUniqueKey` because its NOT a group by at all.  Its a toMap - it was badly named